File: quark/test/emit/macro_eval/py/macro_eval.py
from quark_runtime import *

class side_effects(object):
    def _init(self):
        self.trace = None

    def __init__(self):
        self._init()
        self.trace = u"T"

    def R(self):
        (self).trace = ((self).trace) + (u"R")
        return (self).trace

    def A(self):
        (self).trace = ((self).trace) + (u"A")
        return (self).trace

    def E(self):
        (self).trace = ((self).trace) + (u"CE")
        return (self).trace



def triop(a, b, c):
    if (a):
        return b
    else:
        return c




def main():
    a = side_effects();
    _println((triop(((a).R()) != (u"a"), (a).R(), u"null")) + (triop(((a).A()) != (u"b"), (a).A(), u"null")));
    _println((u"trace: ") + ((a).E()));


if __name__ == "__main__":
    main()

File: quark/test/emit/macro_eval/js/macro_eval.js
var _qrt = require("quark_runtime.js");

// CLASS side_effects

function side_effects() {
    this.__init_fields__();
    this.trace = "T";
}
exports.side_effects = side_effects;

function side_effects__init_fields__() {
    this.trace = null;
}
side_effects.prototype.__init_fields__ = side_effects__init_fields__;

function side_effects_R() {
    (this).trace = ((this).trace) + ("R");
    return (this).trace;
}
side_effects.prototype.R = side_effects_R;

function side_effects_A() {
    (this).trace = ((this).trace) + ("A");
    return (this).trace;
}
side_effects.prototype.A = side_effects_A;

function side_effects_E() {
    (this).trace = ((this).trace) + ("CE");
    return (this).trace;
}
side_effects.prototype.E = side_effects_E;

function triop(a, b, c) {
    if (a) {
        return b;
    } else {
        return c;
    }
}
exports.triop = triop;


function main() {
    var a = new side_effects();
    _qrt.print((triop(((a).R()) !== ("a"), (a).R(), "null")) + (triop(((a).A()) !== ("b"), (a).A(), "null")));
    _qrt.print(("trace: ") + ((a).E()));
}
exports.main = main;

main();
